catalog: disentangle type hydration logic from descriptors

postamar opened the pull request:
This PR disentangles the `types.T` hydration logic from the `catalog.Descriptor` interfaces themselves. This is a prerequisite to disentangling the lifecycle of the `types.T` metadata themselves.

Informs #93314.

Release note: None

Commit:
catalog: rename hydrateddesc to hydrateddesccache

This commit only renames things and moves code around. The
hydrateddesccache package now contains only the parts of hydrateddesc
which pertain to the cache data structure, the rest of the hydration
logic has been moved to descs because it is only ever used by the
Collection.

Release note: None

Commit:
catid,sql: remove errors from OID to descriptor ID converters

Instead, return zero when the OID is not for a user-defined object.
This simplifies the code with no practical loss of safety.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94246)
<!-- Reviewable:end -->

Commit:
typedesc: remove HydrateTypeInfoWithName, simplify EnsureTypeIsHydrated

This refactor simplifies the hydration logic owned by type descriptors
without changing it in any significant way.

Release note: None

Commit:
typedesc: remove MakeTypesT, add AsTypesT

This refactor moves all remaining hydration logic from inside the
type descriptor implementations elsewhere in the typedesc package.

Release note: None

Commit:
catalog: unify hydration logic

This commit unifies the hydration logic which previously was specialized
by descriptor subtype.

Release note: None

Commit:
typedesc: lazy hydration of implicit recort types

Release note: None

Commit:
sql: remove TestTypeSchemaChangeHandlesDeletedDescriptor

This commit removes a test of dubious value: type descriptors do not get
deleted without going through the DROP state first and removing
forwards backwards references first. In any case, outright deletion
would not be safe with regards to type hydration.

Release note: Nonae

postamar commented:
This is ready for review now and is best reviewed commit-by-commit.

fqazi reviewed (approved):
Nice work, looks a lot cleaner and more logical. That said I'm not super familiar with this area, but didn't spot anything obvious.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 12 of 12 files at r1, 28 of 28 files at r2, 5 of 5 files at r3, 17 of 17 files at r4, 25 of 25 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94246#-:-NKtTGH71mW4XCCz2AZ3:b5dzfv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[`-- commits` line 45 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/94246#-NKtSOFmDDGnjjQ2ywTL):*
*record


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/catalog/dbdesc/database_desc.go:
```diff
@@ -532,9 +533,9 @@ func (desc *immutable) GetRawBytesInStorage() []byte {
 	return desc.rawBytesInStorage
 }
 
-// GetRawBytesInStorage implements the catalog.Descriptor interface.
```
good removal

chengxiong-ruan reviewed (approved):
LGTM, very nice work! nothing to complain.

postamar commented on pkg/sql/catalog/typedesc/hydrate.go:
```diff
@@ -0,0 +1,145 @@
+			tm.EnumData.PhysicalRepresentations[i] = desc.GetMemberPhysicalRepresentation(i)
+			tm.EnumData.IsMemberReadOnly[i] = desc.IsMemberReadOnly(i)
+		}
+	case descpb.TypeDescriptor_TABLE_IMPLICIT_RECORD_TYPE:
```
No, we ignore them because there's nothing to do in those cases beyond setting the name and the version. When this function is called all the types.T that this types.T depends on have already been hydrated.

postamar commented:
Thanks for the reviews!

bors r+

postamar commented:
bors r+
